controllers/gateway_controller.go

```diff
@@ -476,5 +447,33 @@ func UpdateGWListenerStatus(ctx context.Context, k8sclient client.Client, gw *ga
 	} else {
 		return fmt.Errorf("no valid listeners for %s", gw.Name)
 	}
+}
 
+func listenerRouteGroupKindSupported(listener gateway_api.Listener) (bool, []gateway_api.RouteGroupKind) {
+	validRoute := true
+	supportedKinds := make([]gateway_api.RouteGroupKind, 0)
+
+	for _, routeGroupKind := range listener.AllowedRoutes.Kinds {
+		if routeGroupKind.Kind == "HTTPRoute" {
+			supportedKinds = append(supportedKinds, gateway_api.RouteGroupKind{
+				Kind: "HTTPRoute",
+			})
+		} else if routeGroupKind.Kind == "GRPCRoute" {
+			if listener.Protocol == gateway_api.HTTPSProtocolType {
+				supportedKinds = append(supportedKinds, gateway_api.RouteGroupKind{
+					Kind: "GRPCRoute",
+				})
+			} else {
+				validRoute = false
+			}
+		} else {
+			validRoute = false
+		}
+	}
+
+	if validRoute {
+		return true, supportedKinds
+	} else {
+		return false, supportedKinds
+	}
```

Comment:
- Guessing from code expected behavior - if list is empty or contains unsupported type return false, am I right? It would be helpful if you document intent in the function description
- Please add unit test
- nit: return arguments order: result comes first and bool-ok second (flip them)

Reply:
From my understanding, this has no impact on controller functionality, and is only for correctness of some metadata (based on Doyoon's description). For the sake of time, let's skip this one for now.